### PR TITLE
Draft. Notify one (not all) for job_finished in ThreadPool

### PR DESCRIPTION
**Pull request by ilejn**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Reduce overhead if many jobs are queued in thread pool awaiting for a thread to execute

- Motivation: Improve behavior if ThreadPool is a bottleneck

### CODE REVIEW

**Comment by ilejn** [comment]
> Great. It seems to be strictly better than before, so we should not see any degradation. But I'm not sure if we have any good tests for overloaded ThreadPool though.
> 
> I wonder what is the rate of jobs per second in your case if the ThreadPool is indeed a bottleneck?

Thanks for your high opinion, although right now unit tests are not passed, so there is a room for improvement ;)
I have a perf tests but not sure if it can be a part of unit tests.

As far as I understand, in our customers' setup there are several (not dozens) thousands threads that try to use a ThreadPool. I know neither size or nature of the thread pool.

**Comment by filimonov** [comment]
> Great. It seems to be strictly better than before, so we should not see any degradation. But I'm not sure if we have any good tests for overloaded ThreadPool though.
> 
> I wonder what is the rate of jobs per second in your case if the ThreadPool is indeed a bottleneck?

We do occasionally observe on systems under heavy load (with hundreds of concurrent queries and thousands of active threads) that CPU utilization remains low. In these cases, ClickHouse tends to process even simple queries like `SELECT` 1 or `/ping` requests very slowly (like 5 seconds). At the same time, other processes on the same node don't seem to suffer as much from the high thread count.

A significant number of traces (found in trace_log, random checks of stack traces, or by examining the process with gdb) indicate a problem with ThreadPool: sometimes even hundreds of threads get locked in scheduleImpl, which perfectly explains the slow performance. Therefore, we're investigating potential issues. Our current suspects are excessive log contention due to notify_all, and the slowness of thread creation/destruction (which also occurs under the same lock).

Interestingly, the problem seems more visible in Docker/Kubernetes environments.

Here is an example stacktrace: https://gist.github.com/filimonov/ecaeae89405d24d449d54963d17abbaf 

It was collected during a rather basic and very synthetic test, which nonetheless helps to highlight the issue (at least on my laptop in Docker):

```
clickhouse-client --query="select sum(sleepEachRow(0.000001)) from numbers_mt(200000000) settings max_threads=4500, max_block_size = 1" &> /dev/null &
```

In real-world scenarios, we usually see a few hundred concurrent queries, each running several threads.

**Comment by filimonov** [comment]
Some simple benchmarking attempt:

the test:
```
time clickhouse-client --query="select sum(sleepEachRow(0.000001)) from numbers_mt(8000000) settings max_threads=4500, max_block_size = 1" &> /dev/null &
sleep 1
clickhouse-benchmark --query='SELECT 1' --query_profiler_real_time_period_ns=10000000 -t 20
```

before 
```
Queries executed: 8.

localhost:9000, queries: 8, QPS: 0.346, RPS: 0.346, MiB/s: 0.000, result RPS: 0.346, result MiB/s: 0.000.

0.000%		0.358 sec.	
10.000%		0.944 sec.	
20.000%		0.944 sec.	
30.000%		1.981 sec.	
40.000%		1.998 sec.	
50.000%		2.281 sec.	
60.000%		2.281 sec.	
70.000%		2.549 sec.	
80.000%		4.125 sec.	
90.000%		4.125 sec.	
95.000%		6.134 sec.	
99.000%		6.134 sec.	
99.900%		6.134 sec.	
99.990%		6.134 sec.	


root@9b63e3e042c0:/# 
real	0m25.098s
user	0m0.065s
sys	0m0.025s
```


after 
```
localhost:9000, queries: 12, QPS: 0.552, RPS: 0.552, MiB/s: 0.000, result RPS: 0.552, result MiB/s: 0.000.

0.000%		0.016 sec.	
10.000%		0.018 sec.	
20.000%		0.109 sec.	
30.000%		0.469 sec.	
40.000%		1.437 sec.	
50.000%		1.448 sec.	
60.000%		2.459 sec.	
70.000%		2.762 sec.	
80.000%		3.145 sec.	
90.000%		3.671 sec.	
95.000%		3.671 sec.	
99.000%		3.886 sec.	
99.900%		3.886 sec.	
99.990%		3.886 sec.	


root@9b63e3e042c0:/# 
real	0m23.401s
user	0m0.040s
sys	0m0.041s
```

That sounds like the main issue was not solved by that PR (while maybe there was a small improvement that should be confirmed in lab env)

**Comment by filimonov** [comment]
See also https://github.com/ClickHouse/ClickHouse/pull/58976

**Comment by serxa** [comment]
> Here is an example stacktrace: https://gist.github.com/filimonov/ecaeae89405d24d449d54963d17abbaf

@filimonov 
Given stacktraces it looks like `ScheduleImpl` is just referenced in every frame as a place where some lambda was created. And looks like `ReadProgressCallback::onProgress` is the real problem. All threads are blocked by waiting on `limits_and_quotas_mutex`. I made the scope of it smaller and renamed it [here](https://github.com/ClickHouse/ClickHouse/commit/f3c12b854386f9c57fe5f8a40193eaae3471a9f8), but it existed even before. And I wonder if it is required at all. At the first glance, quotas look like to be thread-safe, but I'm not  sure about limits.

cc @nickitat You may be also interested in this strange synchronization point of all threads.

**Comment by serxa** [comment]
You should also check if this fix https://github.com/ClickHouse/ClickHouse/pull/56105 is included for real workload.

**Comment by serxa** [comment]
Anyways, this PR is good to have regardless

**Comment by nickitat** [comment]
I have no objections for the proposed patch, but all this guesses look really questionable. would be better to have a proper reproducer and analyse it carefully.

**Comment by ilejn** [comment]
> I have no objections for the proposed patch, but all these guesses look really questionable. would be better to have a proper reproducer and analyse everything carefully.

@nickitat  Yes, we are currently analyzing.
The effect of the patch is not significant.

BTW, we are seeing very slow thread creation in CH sometimes. E.g. > 1 sec.
Cannot explain it by number of existing threads or overload.
Any thoughts may be?

**Comment by nickitat** [comment]
> > I have no objections for the proposed patch, but all these guesses look really questionable. would be better to have a proper reproducer and analyse everything carefully.
> 
> @nickitat Yes, we are currently analyzing. The effect of the patch is not significant.
> 
> BTW, we are seeing very slow thread creation in CH sometimes. E.g. > 1 sec. Cannot explain it by number of existing threads or overload. Any thoughts may be?

what is thread creation exactly and how do you measure its latency?

**Comment by ilejn** [comment]
> what is thread creation exactly and how do you measure its latency?

Looks like it is actual syscall time.
Currently we measure it by timers or suchlike. Fighting with perf, thinking about eBPF.

**Comment by nickitat** [comment]
> > what is thread creation exactly and how do you measure its latency?
> 
> Looks like it is actual syscall time. Currently we measure it by timers or suchlike. Fighting with perf, thinking about eBPF.

I think the best approach is to try to narrow the scope with additional CH-level metrics. seems like you already tried to do it here #58976.

**Comment by azat** [comment]
>All threads are blocked by waiting on limits_and_quotas_mutex.

This is indeed what I observed as well - https://github.com/ClickHouse/ClickHouse/pull/58879

**Comment by ilejn** [comment]
> > All threads are blocked by waiting on limits_and_quotas_mutex.
> 
> This is indeed what I observed as well - #58879

As far as I can see, what we are facing is not about limits_and_quotas_mutex or not only about it.

File: src/Common/ThreadPool.cpp [hunk 1]
```diff
@@ -211,16 +215,26 @@ ReturnType ThreadPoolImpl<Thread>::scheduleImpl(Job job, Priority priority, std:
...
+
+            ++scheduled_jobs;
+            scheduled_jobs_incremented = true;
+            lock.unlock();
+
 
             try
             {
-                threads.front() = Thread([this, it = threads.begin()] { worker(it); });
+                *thread_it = Thread([this, thread_it] { worker(thread_it); });
```
Comment (filimonov) [inline src/Common/ThreadPool.cpp:227]: now i remember why I needed that interator holder - if you pass the iterator out of the lock (used to sync access to the `threads` list), the iterator can become incorrect / invalid if the `threads` will be modified by some different thread (for example somebody else will do clear).

**Comment by alexey-milovidov** [comment]
```
Queries executed: 12696.

localhost:9000, queries: 12696, QPS: 634.774, RPS: 634.774, MiB/s: 0.001, result RPS: 634.774, result MiB/s: 0.001.

0.000%          0.000 sec.
10.000%         0.000 sec.
20.000%         0.000 sec.
30.000%         0.000 sec.
40.000%         0.000 sec.
50.000%         0.000 sec.
60.000%         0.000 sec.
70.000%         0.000 sec.
80.000%         0.000 sec.
90.000%         0.001 sec.
95.000%         0.001 sec.
99.000%         0.001 sec.
99.900%         0.042 sec.
99.990%         1.032 sec.
```

**Comment by alexey-milovidov** [comment]
I like this change, but we need to prove it again and resubmit.

**Pull request closed without merging**